### PR TITLE
Add InfExtendedTime for TimeType support

### LINKED ISSUE #3: Support Date, DateTime
Hi,
First off, thanks for making this package! 

I am wondering if you would be open to adding Date and DateTime support to this package, as I would like to use Infinity as an unbounded date for date intervals.
I guess we would need to expand on the `InfExtended` datatype, or even make something else such as `InfDateExtended`.

Would you be open to taking a PR that adds this support, as I would love to start working on such a thing right away. 

Thanks!

### PR DESCRIPTION
**Pull request by fchorney**
Alright here's a first pass at attempting to add TimeType support. 
Tried to take #4 and #5 into account. 

Closes #3

### CODE REVIEW

File: src/infextendedtime/comparison.jl [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Base.isfinite(x::InfExtendedTime) = x.flag == finite && isfinite(x.finitevalue)
```
Comment (fchorney) [inline src/infextendedtime/comparison.jl:1]: I don't know that I really need that `isfinite(x.finitevalue)` to be there.

File: src/infextendedtime/comparison.jl [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+Base.isinf(x::InfExtendedTime) = isposinf(x) || isneginf(x)
+Base.:(==)(x::InfExtendedTime, y::InfExtendedTime) = (isfinite(x) && isfinite(y)) ? x.finitevalue == y.finitevalue : x.flag == y.flag
+Base.hash(x::InfExtendedTime, h::UInt) = isfinite(x) ? hash(x.finitevalue, h) : hash(x.flag, h)
```
Comment (omus) [inline src/infextendedtime/comparison.jl:4]: Currently wouldn't support hashing different infinities (e.g. `InfExtended` vs. `InfExtendedTime`) to the same result. I'm not sure what the desired behaviour is.

File: src/infextendedtime/comparison.jl [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+function Base.:<(x::InfExtendedTime, y::InfExtendedTime)
```
Comment (omus) [inline src/infextendedtime/comparison.jl:6]: Normally you just overload `isless`

**Review by omus — COMMENTED** [review]
I suspect most of this code was copied from `InfExtended`. It would probably be best to do some code generation rather than copy/paste. For example:
```julia
for InfExt in (InfExtended, InfExtendedTime)
    @eval begin 
        Base.promote_rule(::Type{$InfExt{T}}, ::Type{$InfExt{S}}) where {T, S} = $InfExt(promote_type(T, S))
        Base.promote_rule(::Type{$InfExt{T}}, ::Type{S}) where {T, S} = $InfExt(promote_type(T, S))
        Base.promote_rule(::Type{$InfExt{T}}, ::Type{Infinite}) where T = $InfExt{T}
    end
end
```

File: src/infextendedtime/comparison.jl (hunk 2 above)
Comment (fchorney) [reply src/infextendedtime/comparison.jl:4]: Good point.

**Comment by fchorney** [comment]
> I suspect most of this code was copied from `InfExtended`. It would probably be best to do some code generation rather than copy/paste. For example:
> 
> ```julia
> for InfExt in (InfExtended, InfExtendedTime)
>     @eval begin 
>         Base.promote_rule(::Type{$InfExt{T}}, ::Type{$InfExt{S}}) where {T, S} = $InfExt(promote_type(T, S))
>         Base.promote_rule(::Type{$InfExt{T}}, ::Type{S}) where {T, S} = $InfExt(promote_type(T, S))
>         Base.promote_rule(::Type{$InfExt{T}}, ::Type{Infinite}) where T = $InfExt{T}
>     end
> end
> ```

I guess this kind of breaks down the separated structure that #5 was looking for. Though I guess a few things could be moved into a more generic section.

**Comment by omus** [comment]
>  I'm thinking maybe this could be part of #5?

Very true, maybe then it would good to add comments as to parts that were copied from `InfExtended` but had the type changed to `InfExtendedTime`? Would make it easier to catch mistakes

**Comment by cjdoris** [comment]
Thanks for this, it looks really good. Please merge in the master branch --- I renamed the (badly thought through) `InfFlag` constants to be uppercase (e.g. `POSINF`), because they conflicted with things from `Infinity.Utils`, which will break your code somewhere.

**Comment by cjdoris** [comment]
> > I'm thinking maybe this could be part of #5?
> 
> Very true, maybe then it would good to add comments as to parts that were copied from `InfExtended` but had the type changed to `InfExtendedTime`? Would make it easier to catch mistakes

If there is code that is sensible to generate together for both `InfExtended` (which I'm considering renaming to `InfExtendedReal`) and `InfExtendedTime` then please do. Maybe there should be directories `infextendedcommon`, `infextendedreal` and `infextendedtime`.

File: src/infextendedtime/comparison.jl (hunk 2 above)
Comment (cjdoris) [reply src/infextendedtime/comparison.jl]: I'd like the hash to simply defer to whatever the true value is, i.e. the `finitevalue` or `PosInf` or `NegInf`. It's fine that the infinite date has the same hash as the infinite number.

**Pull request merged**